jerry-core/lit/lit-magic-strings.c

```diff
@@ -145,22 +145,45 @@ lit_is_utf8_string_magic (const lit_utf8_byte_t *string_p, /**< utf-8 string */
                           lit_utf8_size_t string_size, /**< string size in bytes */
                           lit_magic_string_id_t *out_id_p) /**< [out] magic string's id */
 {
-  /* TODO: Improve performance of search */
+  static const lit_utf8_byte_t * const magic_strings[] JERRY_CONST_DATA =
+  {
+#define LIT_MAGIC_STRING_DEF(id, utf8_string) \
+    (const lit_utf8_byte_t *) utf8_string,
+#include "lit-magic-strings.inc.h"
+#undef LIT_MAGIC_STRING_DEF
+  };
 
-  for (lit_magic_string_id_t id = (lit_magic_string_id_t) 0;
-       id < LIT_MAGIC_STRING__COUNT;
-       id = (lit_magic_string_id_t) (id + 1))
+  *out_id_p = LIT_MAGIC_STRING__COUNT;
+
+  int first = 0; /**< First magic string element */
+  int last = (int) (LIT_MAGIC_STRING__COUNT - 1); /**< Last magic string element */
+
+  while (first <= last)
   {
-    if (lit_compare_utf8_string_and_magic_string (string_p, string_size, id))
+    int middle = ((first + last) / 2); /**< mid point of search */
+
+    int compare = memcmp (magic_strings[middle], string_p, string_size);
```

Comment:
I just realized that this is unsafe if the input string contains \0 which is allowed.